File: backend/app/schemas/documents.py
# =============================================================================
# File: documents.py
# Module/Service: Document Ingestion Service
# Layer: Schema
# Purpose: Pydantic models for Documents / Versions / Pipeline status (FR2).
# Responsibilities:
#   - Match OpenAPI Document, DocumentVersion, DocumentListResponse, PipelineRun
# Dependencies:
#   - Pydantic, app.models.enums
# Public Exports:
#   - DocumentResponse, DocumentListResponse, DocumentVersionResponse
#   - PipelineStageLogResponse, PipelineRunResponse
# Database/Table: documents, document_versions, pipeline_runs, pipeline_stage_logs
# Related Modules: app.api.documents, docs/Enterprise_notebooklm_openapi.yaml
# Important Notes: storage_path is internal — not exposed in DocumentVersion schema.
# =============================================================================

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DocumentResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    title: str
    file_type: Literal["pdf", "docx", "xlsx", "pptx", "txt"]
    current_version_id: UUID | None = None
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]
    page: int
    page_size: int
    total: int


class DocumentVersionResponse(BaseModel):
    id: UUID
    document_id: UUID
    uploaded_by: UUID
    version_number: int
    file_size_bytes: int
    checksum_sha256: str
    page_count: int | None = Field(
        default=None,
        description=(
            "Physical page/slide/sheet count for PDF/PPTX/XLSX. "
            "For DOCX: logical section count (by heading), not printed Word pages."
        ),
    )
    status: Literal["processing", "ready", "failed"]
    is_current: bool
    created_at: datetime
    preview_status: Literal["pending", "processing", "completed", "failed"] = "pending"
    preview_type: Literal["pdf", "html", "image"] | None = None
    preview_generated_at: datetime | None = None


class PipelineStageLogResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: UUID
    stage: Literal[
        "ocr_cleaning",
        "chunking",
        "preview_generation",
        "document_understanding",
        "cleaning_normalize",
        "hierarchical_chunking",
        "embedding",
        "graph_extraction",
        "indexing",
    ]
    status: Literal["pending", "running", "completed", "failed"]
    duration_ms: int | None = None
    metadata: dict[str, Any] | None = Field(default=None, alias="metadata")
    error_message: str | None = None


class PipelineRunResponse(BaseModel):
    id: UUID
    document_version_id: UUID
    status: Literal["pending", "running", "completed", "failed"]
    retry_count: int
    error_message: str | None = None
    stages: list[PipelineStageLogResponse] = Field(default_factory=list)
    started_at: datetime | None = None
    completed_at: datetime | None = None
    # Optional document context — populated on admin list (JOIN already required
    # for workspace scope). Null on endpoints that only load the run row.
    document_id: UUID | None = None
    document_title: str | None = None
    file_type: Literal["pdf", "docx", "xlsx", "pptx", "txt"] | None = None
    version_number: int | None = None


class DocumentChunkResponse(BaseModel):
    """Viewer / AI-metadata chunk row (deep-link by chunk_id)."""

    id: UUID
    document_id: UUID
    document_version_id: UUID
    chunk_index: int
    content: str
    page_number: int | None = None
    section_index: int | None = None
    section: str | None = None
    heading_path: str | None = None
    section_path: str | None = Field(
        default=None,
        description="Alias of heading_path for viewer TOC / AI panel.",
    )
    bounding_box: list[float] | None = Field(
        default=None,
        description="Optional [x, y, w, h] from layout_metadata when available.",
    )
    start_offset: int | None = None
    end_offset: int | None = None


class DocumentChunkListResponse(BaseModel):
    document_id: UUID
    document_version_id: UUID | None = None
    document_title: str
    file_type: Literal["pdf", "docx", "xlsx", "pptx", "txt"]
    viewer_kind: Literal["pdf", "original_download"] = "original_download"
    preview_status: Literal["pending", "processing", "completed", "failed"] = "pending"
    preview_type: Literal["pdf", "html", "image"] | None = None
    preview_generated_at: datetime | None = None
    heading_tree: list[dict[str, Any]] = Field(default_factory=list)
    items: list[DocumentChunkResponse] = Field(default_factory=list)

File: frontend/types/documents.ts
/**
 * =============================================================================
 * File: documents.ts
 * Module/Service: Document Ingestion Service (Web App)
 * Layer: UI
 * Purpose: TypeScript types for Document/Version/Pipeline matching OpenAPI (FR2).
 * Responsibilities:
 *   - Align frontend Document/DocumentVersion/PipelineRun with backend Pydantic
 *     schemas (app/schemas/documents.py)
 * Dependencies:
 *   - docs/Enterprise_notebooklm_openapi.yaml
 * Public Exports:
 *   - Document, DocumentListResponse, DocumentVersion, DocumentVersionStatus
 *   - PipelineRun, PipelineStageLog, PipelineStageName, PipelineStatus
 * Database/Table: documents, document_versions, pipeline_runs, pipeline_stage_logs
 * Related Modules: lib/api-client, lib/pipeline-stages, features/documents/*
 * Important Notes: PipelineStageName includes legacy v2 values (ocr_cleaning,
 *   chunking) returned by old runs — v3 UI only renders PIPELINE_STAGE_ORDER (6).
 * =============================================================================
 */

export type FileType = "pdf" | "docx" | "xlsx" | "pptx" | "txt";

export type Document = {
  id: string;
  workspace_id: string;
  title: string;
  file_type: FileType;
  current_version_id: string | null;
  created_at: string;
  updated_at: string;
};

export type DocumentListResponse = {
  items: Document[];
  page: number;
  page_size: number;
  total: number;
};

export type DocumentChunk = {
  id: string;
  document_id: string;
  document_version_id: string;
  chunk_index: number;
  content: string;
  page_number?: number | null;
  section_index?: number | null;
  section?: string | null;
  heading_path?: string | null;
  section_path?: string | null;
  bounding_box?: number[] | null;
  start_offset?: number | null;
  end_offset?: number | null;
};

export type DocumentChunkListResponse = {
  document_id: string;
  document_version_id: string | null;
  document_title: string;
  file_type: FileType;
  viewer_kind?: "pdf" | "original_download";
  preview_status?: PreviewStatus;
  preview_type?: PreviewType | null;
  preview_generated_at?: string | null;
  heading_tree?: Array<Record<string, unknown>>;
  items: DocumentChunk[];
};

export type DocumentVersionStatus = "processing" | "ready" | "failed";

export type PreviewStatus = "pending" | "processing" | "completed" | "failed";
export type PreviewType = "pdf" | "html" | "image";

export type DocumentVersion = {
  id: string;
  document_id: string;
  uploaded_by: string;
  version_number: number;
  file_size_bytes: number;
  checksum_sha256: string;
  page_count: number | null;
  status: DocumentVersionStatus;
  is_current: boolean;
  created_at: string;
  preview_status?: PreviewStatus;
  preview_type?: PreviewType | null;
  preview_generated_at?: string | null;
};

/** v3 stage order including Preview Generation before AI stages. */
export type PipelineStageNameV3 =
  | "preview_generation"
  | "document_understanding"
  | "cleaning_normalize"
  | "hierarchical_chunking"
  | "embedding"
  | "graph_extraction"
  | "indexing";

/** Legacy v2 stages a run can still report; not part of the 6-step v3 tracker UI. */
export type PipelineStageNameLegacy = "ocr_cleaning" | "chunking";

export type PipelineStageName = PipelineStageNameV3 | PipelineStageNameLegacy;

export type PipelineStatus = "pending" | "running" | "completed" | "failed";

export type PipelineStageLog = {
  id: string;
  stage: PipelineStageName;
  status: PipelineStatus;
  duration_ms: number | null;
  metadata: Record<string, unknown> | null;
  error_message: string | null;
};

export type PipelineRun = {
  id: string;
  document_version_id: string;
  status: PipelineStatus;
  retry_count: number;
  error_message: string | null;
  stages: PipelineStageLog[];
  started_at: string | null;
  completed_at: string | null;
  /** Present on admin pipeline-runs list; may be null elsewhere. */
  document_id?: string | null;
  document_title?: string | null;
  file_type?: FileType | null;
  version_number?: number | null;
};
